include/pybind11/pybind11.h

```diff
@@ -1514,6 +1514,153 @@ using default_holder_type = smart_holder;
+                                       detail::type_uses_smart_holder_type_caster<D>,
+                                       detail::none_of<std::is_pointer<D>,
+                                                       detail::is_std_unique_ptr<D>,
+                                                       detail::is_std_shared_ptr<D>>>::value>> {
+
+    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    static cpp_function readonly(PM pm, const handle &hdl) {
+        return cpp_function(
+            [pm](const std::shared_ptr<T> &c_sp)
+                -> std::shared_ptr<typename std::add_const<D>::type> {
+                return std::shared_ptr<typename std::add_const<D>::type>(c_sp, &(c_sp.get()->*pm));
+            },
+            is_method(hdl));
+    }
+
+    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    static cpp_function read(PM pm, const handle &hdl) {
+        return cpp_function(
+            [pm](const std::shared_ptr<T> &c_sp) -> std::shared_ptr<D> {
+                return std::shared_ptr<D>(c_sp, &(c_sp.get()->*pm));
+            },
+            is_method(hdl));
+    }
+
+    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    static cpp_function write(PM pm, const handle &hdl) {
+        return cpp_function([pm](T &c, const D &value) { c.*pm = value; }, is_method(hdl));
+    }
+};
+
+// smart_holder specializations for std::unique_ptr members.
+// read disowns the member unique_ptr.
+// write disowns the passed Python object.
+// readonly is disabled (static_assert) because there is no safe & intuitive way to make the member
+// accessible as a Python object without disowning the member unique_ptr. A .def_readonly disowning
+// the unique_ptr member is deemed highly prone to misunderstandings.
+template <typename T, typename D>
+struct property_cpp_function<
+    T,
+    D,
+    detail::enable_if_t<detail::all_of<
+        detail::type_uses_smart_holder_type_caster<T>,
+        detail::is_std_unique_ptr<D>,
+        detail::type_uses_smart_holder_type_caster<typename D::element_type>>::value>> {
+
+    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    static cpp_function readonly(PM, const handle &) {
+        static_assert(!detail::is_std_unique_ptr<D>::value,
+                      "def_readonly cannot be used for std::unique_ptr members.");
+        return cpp_function{}; // Unreachable.
+    }
+
+    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    static cpp_function read(PM pm, const handle &hdl) {
+        return cpp_function(
+            [pm](const std::shared_ptr<T> &c_sp) -> D { return D{std::move(c_sp.get()->*pm)}; },
+            is_method(hdl));
+    }
+
+    template <typename PM, must_be_member_function_pointer<PM> = 0>
+    static cpp_function write(PM pm, const handle &hdl) {
+        return cpp_function([pm](T &c, D &&value) { c.*pm = std::move(value); }, is_method(hdl));
```

Comment:
Any reason why std::move instead of std::forward?

Reply:
In other words, is this really an rvalue or should it be a universal forwarding reference?

Reply:
Ah, nvm I think I get it because D is always a unique ptr.